scipy/sparse/linalg/_eigen/lobpcg/tests/test_lobpcg.py

```diff
@@ -359,47 +424,63 @@ def test_verbosity():
 def test_tolerance_float32():
     """Check lobpcg for attainable tolerance in float32.
     """
-    rnd = np.random.RandomState(0)
+    rng = np.random.RandomState(0)
     n = 50
     m = 3
     vals = -np.arange(1, n + 1)
     A = diags([vals], [0], (n, n))
     A = A.astype(np.float32)
-    X = rnd.standard_normal((n, m))
+    X = rng.random((n, m))
     X = X.astype(np.float32)
     eigvals, _ = lobpcg(A, X, tol=1.25e-5, maxiter=50, verbosityLevel=0)
     assert_allclose(eigvals, -np.arange(1, 1 + m), atol=2e-5, rtol=1e-5)
 
 
 def test_random_initial_float32():
```

Comment:
I think it would be worthwhile to ensure that all reasonable dtypes get exercised, so we can ensure that `_dot_inplace` has no gaps where people run into issues.

Note: this suggestion is not intended to be committed, I rely on you to know which set of dtypes is relevant.
```suggestion
@pytest.mark.parametrize("dtype", [list-of-relevant-dtypes])
def test_random_initial_dtype(dtype):
```

Reply:
there is already a test `test_diagonal_data_types` for all reasonable dtypes that relies on _dot_inplace, including `float32` that passes. So `float32` by itself is not a problem. 

I could of course exclude `float32` as you propose, if no actual reason if found for the failure in time.

Reply:
That test (`test_diagonal_data_types`) does a whole lot more than templating on the dtype, but does not cover many dtypes. Additionally, since its marked as slow it will also not run in most cases.

Aside from my preference of wanting that test to be heavily parametrized (unrelated to this PR!), I would like to add a simple dtype parametrization to this test. For example, `lobpcg` will happily run through if the matrices are integers or complex. The minimum list of dtypes we should check IMO is:
```
np.float32
np.float64
np.int32
np.uint32
np.int64
np.uint64
np.complex64
np.complex128
```

Reply:
Modifying the already very hard to read `test_diagonal` is not an alternative to parametrizing this test.